Cython/Compiler/Code.py

```diff
@@ -1666,6 +1674,65 @@ def generate_string_constants(self):
                     Naming.stringtab_cname,
                     init_constants.error_goto(self.module_pos)))
 
+    def generate_codeobject_constants(self):
+        # TODO:
+        # There will be a significant optimization here once we've got rid of the
+        # global "#define"s pointing into the module state. When this is done the
+        # references to constants in the code object table can be done using
+        # offsetof, and it can be defined as a C constant global.
+        # For now it must be defined as a function local.
+        self.parts['decls'].putln("static int __Pyx_CreateCodeTabAndInitCode(void); /* proto */")
+
+        w = self.parts['pycodeobj_table']
+        w.putln("static int __Pyx_CreateCodeTabAndInitCode(void) {")
+
+        if not self.codeobject_constants:
+            w.putln("}")
+            return
+
+        self.use_utility_code(UtilityCode.load_cached("InitCodeObjs", "ModuleSetupCode.c"))
+
+        self.parts['module_state'].putln("PyObject *%s[%s];" % (
+                Naming.codeobjtab_cname, len(self.codeobject_constants)))
+
+        self.parts['module_state_clear'].putln("for (int n=0; n<%s; ++n) {" %
+                                               len(self.codeobject_constants))
+        self.parts['module_state_clear'].putln("Py_CLEAR(clear_module_state->%s[n]);" %
+            Naming.codeobjtab_cname)
+        self.parts['module_state_clear'].putln("}")
+        self.parts['module_state_traverse'].putln("for (int n=0; n<%s; ++n) {" %
+                                                  len(self.codeobject_constants))
+        self.parts['module_state_traverse'].putln("Py_VISIT(traverse_module_state->%s[n]);" %
+            Naming.codeobjtab_cname)
+        self.parts['module_state_traverse'].putln("}")
+
+        self.parts['module_state_defines'].putln("#define %s %s->%s" % (
+            Naming.codeobjtab_cname, Naming.modulestateglobal_cname, Naming.codeobjtab_cname
+        ))
+
+        w.putln("__Pyx_CodeObjectTabEntry tab[] = {")
```

Comment:
Didn't you write in the stringtab discussion that local arrays produce a lot of code?
Most of the values are constants, and e.g. for string constants (and tuples), you could store the string's index in the string table (as you do for the file name) and pass the module's string table into the init function, to avoid a dependency on any init-time created pointers. That would make the code object table static data.